enhancements/monitoring/metrics-collection-profiles.md

```diff
@@ -0,0 +1,483 @@
+  rules being deployed, but this is not a good idea because: 
+  - CMO acting behind the back of other operators isn't a good principle (less
+    explicit, more brittle), it would require CMO to watch the components
+    monitors and generates new resources that would be reconciled by
+    prometheus-operator instead of the original ones, this would also translate
+    into a nightmare from a maintainability point of view;
+  - it's likely to be less efficient because the relabeling configs would be
+    very complex and expensive in terms of processing.
+  - Hypershift implements a [similar
+    strategy](https://github.com/openshift/hypershift/pull/1294), but again this
+    would only work for service monitors that CMO deploys directly
+- Let users configure themselves the Prometheus scraping interval. This solution
+  was discarded because:
+  - Users might not be fully aware of the impact that changing this interval
+    might have. Too low of an interval and users might overwhelm Prometheus and
+    exhaust its memory. In contrast, too long of an interval might render the
+    default alerts ineffective;
+  - It's not advisable to increase the scrape interval to more than 2 minutes.
+    Staleness occurs at 5 minutes (by default) which is would cause gaps in
+    graphs/dashboards;
+  - Many upstream dashboards were built while having the 30 second interval in
+    mind;
+  - Scrapes can fail, but the user might not be mindful of this and set a high
+    scrape interval;
+- Add a separate container to prometheus-operator (p-o) that would be used by
+  p-o to modify the prometheus config according to a metrics collection profile.
+  - This container would perform an analysis on what metrics were being used.
+    Then it would provide prometheus operator with this list.
+  - Prometheus-operator with the list provided by this new component would know
+    what scraping targets it could change to keep certain metrics.
+- Recently Azure also added support for metrics collection profiles:
+  - [Azure
+    Docs](https://learn.microsoft.com/en-us/azure/azure-monitor/essentials/prometheus-metrics-scrape-configuration-minimal)
+  -  https://github.com/Azure/prometheus-collector
+  - In their approach they also have
+  [hardcoded](https://github.com/Azure/prometheus-collector/blob/66ed1a5a27781d7e7e3bb1771b11f1da25ffa79c/otelcollector/configmapparser/tomlparser-default-targets-metrics-keep-list.rb#L28)
+  set of metrics that are only consumed when the minimal profile is enabled.
+  However, customers are also able to extend this minimal profile with regexes
+  to include metrics which might be interesting to them.
+- Leverage [installer capabilities](https://docs.google.com/document/d/1I-YT7LKKDHSBLB6Hmg0tZ54DWjrAxlVdXxlViShMu-0/edit#heading=h.848jsje80fru)
+  - After some consideration we decided to abandon this idea since it would only
+    work for resources controlled by CVO which is not the case for the majority
+    of ServiceMonitors.
```

Comment:
We, components owners already rely on installer capabilities for enabling/disabling Console related SM/Rules (e.g https://github.com/openshift/cluster-kube-apiserver-operator/pull/1542/files), so having yet another thing to configure is not very ideal considering that we will still have to maintain our code relying on the Console capability from the installer.

Would it be perhaps possible to unify these approach to make it easier to integrate and maintain for component owners? An option could be to require component owners to let their Prometheus resources be managed by CVO and then leverage installer capabilities.

@bparees could you have a look to this enhancement since you worked on installer capabilities?

Reply:
Interesting. I looked into our code again as well and yes it turns out that there are some rules that are managed by our operator and not CVO and we already have some in-house conditional logic for SNO: https://github.com/openshift/cluster-kube-apiserver-operator/blob/03c5a03d1bd0a952655aa180bd2708802e8cc1a4/pkg/operator/starter.go#L216

I am just afraid that there might be too many ways at the moment to customize the data we expose and that it will be a pain to maintain going forward. Unifying all the existing knobs for monitoring under collection profiles might also be a good way to simplify the existing processes.

Reply:
in addition to what @simonpasquier said, i think caps are not a good fit for this because it's easy to envision someone wanting a cap (like image-registry) but then wanting to choose the level of metrics that cap should provide (minimal vs full).  There might be ways to solve this with judicious use of cap-names on ServiceMonitor resources and the introduction of new caps, but i think it would be pretty ugly